File: java-backend/src/main/java/com/forge/auth/service/RiskService.java
package com.forge.auth.service;

import com.forge.auth.entity.CompanyEntity;
import com.forge.auth.entity.RiskScoreEntity;
import com.forge.auth.entity.UserEntity;
import com.forge.auth.repository.AuditLogRepository;
import com.forge.auth.repository.CompanyRepository;
import com.forge.auth.repository.LoginHistoryRepository;
import com.forge.auth.repository.RiskScoreRepository;
import com.forge.auth.repository.SessionRepository;
import com.forge.auth.repository.UserRepository;
import com.forge.shared.api.ApiException;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

@Service
public class RiskService {

    private static final List<String> ALERT_LEVELS = List.of("high", "critical");

    private final RiskScoreRepository riskRepo;
    private final UserRepository userRepo;
    private final CompanyRepository companyRepo;
    private final SessionRepository sessionRepo;
    private final LoginHistoryRepository loginHistoryRepo;
    private final AuditLogRepository auditLogRepo;

    public RiskService(RiskScoreRepository riskRepo,
                       UserRepository userRepo,
                       CompanyRepository companyRepo,
                       SessionRepository sessionRepo,
                       LoginHistoryRepository loginHistoryRepo,
                       AuditLogRepository auditLogRepo) {
        this.riskRepo = riskRepo;
        this.userRepo = userRepo;
        this.companyRepo = companyRepo;
        this.sessionRepo = sessionRepo;
        this.loginHistoryRepo = loginHistoryRepo;
        this.auditLogRepo = auditLogRepo;
    }

    // ═══════════════════════════════════════════════════════════════════════
    // GET ALL RISK SCORES
    // ═══════════════════════════════════════════════════════════════════════

    public List<Map<String, Object>> getAllRiskScores(UUID companyId) {
        List<RiskScoreEntity> scores = companyId == null
                ? riskRepo.findAllByOrderByScoreDesc()
                : riskRepo.findAllByCompanyIdOrderByScoreDesc(companyId);
        return scores.stream().map(this::enrichRiskScore).toList();
    }

    // ═══════════════════════════════════════════════════════════════════════
    // GET ALERTS (high + critical)
    // ═══════════════════════════════════════════════════════════════════════

    public List<Map<String, Object>> getAlerts(UUID companyId) {
        List<RiskScoreEntity> alerts = companyId == null
                ? riskRepo.findAllByLevelInOrderByScoreDesc(ALERT_LEVELS)
                : riskRepo.findAllByCompanyIdAndLevelInOrderByScoreDesc(companyId, ALERT_LEVELS);
        return alerts.stream().map(this::enrichRiskScore).toList();
    }

    // ═══════════════════════════════════════════════════════════════════════
    // CALCULATE COMPANY RISK
    // ═══════════════════════════════════════════════════════════════════════

    @Transactional
    public Map<String, Object> calculateCompanyRisk(UUID companyId) {
        List<Map<String, Object>> factors = new ArrayList<>();
        int totalScore = 0;
        Instant now = Instant.now();
        Instant yesterday = now.minus(24, ChronoUnit.HOURS);

        // 1. Failed logins in last 24h for this company
        long failedLogins = loginHistoryRepo.countByCompanyIdAndStatusAndCreatedAtAfter(
                companyId, "failed", yesterday);
        if (failedLogins > 10) {
            int weight = (int) Math.min(30, failedLogins * 2);
            factors.add(factor("failed_logins", weight, failedLogins + " failed logins in 24h"));
            totalScore += weight;
        }

        // 2. Inactive users ratio
        long totalUsers = userRepo.countByCompanyId(companyId);
        long inactiveUsers = userRepo.countByCompanyIdAndIsActiveFalse(companyId);
        if (totalUsers > 0) {
            double ratio = (double) inactiveUsers / totalUsers;
            if (ratio > 0.5) {
                int weight = (int) Math.round(ratio * 20);
                factors.add(factor("high_inactive_ratio", weight,
                        inactiveUsers + "/" + totalUsers + " users inactive"));
                totalScore += weight;
            }
        }

        // 3. Locked accounts
        long lockedUsers = userRepo.countByCompanyIdAndLockedUntilAfter(companyId, now);
        if (lockedUsers > 0) {
            int weight = (int) Math.min(lockedUsers * 10, 25);
            factors.add(factor("locked_accounts", weight,
                    lockedUsers + " accounts currently locked"));
            totalScore += weight;
        }

        // 4. No admin activity in 30 days
        companyRepo.findById(companyId).ifPresent(company -> {
            // handled separately below since we need to mutate totalScore
        });
        CompanyEntity company = companyRepo.findById(companyId).orElse(null);
        if (company != null && company.getLastActivityAt() != null) {
            long daysSince = ChronoUnit.DAYS.between(company.getLastActivityAt(), now);
            if (daysSince > 30) {
                factors.add(factor("admin_inactivity", 15,
                        "No admin activity for " + daysSince + " days"));
                totalScore += 15;
            }
        }

        // 5. Mass data operations (>20 export audit logs in 24h)
        long massOps = auditLogRepo.countByCompanyIdAndActionContainingIgnoreCaseAndCreatedAtAfter(
                companyId, "export", yesterday);
        if (massOps > 20) {
            factors.add(factor("mass_data_operations", 20,
                    massOps + " export operations in 24h"));
            totalScore += 20;
        }

        totalScore = Math.min(totalScore, 100);
        String level = scoreToLevel(totalScore);

        upsertRiskScore("company", companyId, companyId, totalScore, level, factors);

        Map<String, Object> result = new LinkedHashMap<>();
        result.put("score", totalScore);
        result.put("level", level);
        result.put("factors", factors);
        return result;
    }

    // ═══════════════════════════════════════════════════════════════════════
    // CALCULATE USER RISK
    // ═══════════════════════════════════════════════════════════════════════

    @Transactional
    public Map<String, Object> calculateUserRisk(UUID userId) {
        UserEntity user = userRepo.findById(userId)
                .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "User not found"));

        List<Map<String, Object>> factors = new ArrayList<>();
        int totalScore = 0;
        Instant now = Instant.now();

        // 1. Failed login attempts
        int failedAttempts = user.getFailedLoginAttempts() != null ? user.getFailedLoginAttempts() : 0;
        if (failedAttempts > 3) {
            int weight = Math.min(30, failedAttempts * 5);
            factors.add(factor("failed_logins", weight,
                    failedAttempts + " failed attempts"));
            totalScore += weight;
        }

        // 2. Account locked
        if (user.getLockedUntil() != null && user.getLockedUntil().isAfter(now)) {
            factors.add(factor("account_locked", 20, "Account is currently locked"));
            totalScore += 20;
        }

        // 3. Multiple active sessions (> 5)
        long activeSessions = sessionRepo.countByUserIdAndIsActiveTrueAndExpiresAtAfter(userId, now);
        if (activeSessions > 5) {
            factors.add(factor("many_sessions", 15,
                    activeSessions + " active sessions"));
            totalScore += 15;
        }

        // 4. Admin without 2FA
        boolean isAdmin = "main_admin".equals(user.getRole()) || "admin".equals(user.getRole());
        boolean hasTwoFactor = Boolean.TRUE.equals(user.getTwoFactorEnabled());
        if (isAdmin && !hasTwoFactor) {
            factors.add(factor("no_2fa", 10, "Admin without 2FA"));
            totalScore += 10;
        }

        // 5. No login in >60 days
        if (user.getLastLogin() != null) {
            long daysSince = ChronoUnit.DAYS.between(user.getLastLogin(), now);
            if (daysSince > 60) {
                factors.add(factor("long_inactive", 15,
                        "No login for " + daysSince + " days"));
                totalScore += 15;
            }
        }

        totalScore = Math.min(totalScore, 100);
        String level = scoreToLevel(totalScore);

        upsertRiskScore("user", userId, user.getCompanyId(), totalScore, level, factors);

        Map<String, Object> result = new LinkedHashMap<>();
        result.put("score", totalScore);
        result.put("level", level);
        result.put("factors", factors);
        return result;
    }

    // ═══════════════════════════════════════════════════════════════════════
    // RECALCULATE ALL (admin only)
    // ═══════════════════════════════════════════════════════════════════════

    @Transactional
    public List<Map<String, Object>> recalculateAll() {
        List<CompanyEntity> companies = companyRepo.findByIsActiveTrue();
        List<Map<String, Object>> results = new ArrayList<>();
        for (CompanyEntity c : companies) {
            try {
                Map<String, Object> risk = calculateCompanyRisk(c.getId());
                Map<String, Object> row = new LinkedHashMap<>(risk);
                row.put("company_id", c.getId());
                results.add(row);
            } catch (Exception e) {
                Map<String, Object> row = new LinkedHashMap<>();
                row.put("company_id", c.getId());
                row.put("error", e.getMessage());
                results.add(row);
            }
        }
        return results;
    }

    // ═══════════════════════════════════════════════════════════════════════
    // HELPERS
    // ═══════════════════════════════════════════════════════════════════════

    private void upsertRiskScore(String entityType, UUID entityId, UUID companyId,
                                  int score, String level, List<Map<String, Object>> factors) {
        Instant now = Instant.now();
        RiskScoreEntity entity = riskRepo.findByEntityTypeAndEntityId(entityType, entityId)
                .orElseGet(() -> {
                    RiskScoreEntity n = new RiskScoreEntity();
                    n.setId(UUID.randomUUID());
                    n.setEntityType(entityType);
                    n.setEntityId(entityId);
                    n.setCreatedAt(now);
                    return n;
                });
        entity.setScore(score);
        entity.setLevel(level);
        entity.setFactors(factors);
        entity.setLastCalculatedAt(now);
        entity.setCompanyId(companyId);
        entity.setUpdatedAt(now);
        riskRepo.save(entity);
    }

    private Map<String, Object> enrichRiskScore(RiskScoreEntity r) {
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("id", r.getId());
        m.put("entity_type", r.getEntityType());
        m.put("entity_id", r.getEntityId());
        m.put("score", r.getScore());
        m.put("level", r.getLevel());
        m.put("factors", buildFactorOutput(r.getFactors()));
        m.put("last_calculated_at", r.getLastCalculatedAt());
        m.put("company_id", r.getCompanyId());

        if ("company".equals(r.getEntityType())) {
            companyRepo.findById(r.getEntityId()).ifPresent(c -> {
                Map<String, Object> co = new LinkedHashMap<>();
                co.put("name", c.getName());
                m.put("company", co);
            });
        } else if ("user".equals(r.getEntityType())) {
            userRepo.findById(r.getEntityId()).ifPresent(u -> {
                Map<String, Object> uo = new LinkedHashMap<>();
                uo.put("name", u.getName());
                uo.put("email", u.getEmail());
                m.put("user", uo);
            });
        }
        return m;
    }

    /**
     * Translate stored factor maps into the frontend-expected shape:
     * { name, score, weight, description }
     * Node stores factors as { factor, weight, detail } — we map those fields.
     */
    private List<Map<String, Object>> buildFactorOutput(List<Map<String, Object>> raw) {
        if (raw == null) return List.of();
        return raw.stream().map(f -> {
            Map<String, Object> out = new LinkedHashMap<>();
            String name = f.getOrDefault("factor", "unknown").toString();
            Object w = f.getOrDefault("weight", 0);
            int weight = w instanceof Number n ? n.intValue() : 0;
            out.put("name", name);
            out.put("score", weight);
            out.put("weight", weight);
            out.put("description", f.getOrDefault("detail", name));
            return out;
        }).toList();
    }

    private Map<String, Object> factor(String name, int weight, String detail) {
        Map<String, Object> f = new LinkedHashMap<>();
        f.put("factor", name);
        f.put("weight", weight);
        f.put("detail", detail);
        return f;
    }

    private String scoreToLevel(int score) {
        if (score >= 75) return "critical";
        if (score >= 50) return "high";
        if (score >= 25) return "medium";
        return "low";
    }
}

File: backend/src/services/riskService.js
const { Op, fn, col, literal } = require('sequelize');
const { RiskScore, User, Company, LoginHistory, Session, AuditLog, sequelize } = require('../models');

class RiskService {
  /**
   * Calculate risk score for a company
   */
  async calculateCompanyRisk(companyId) {
    const factors = [];
    let totalScore = 0;

    // 1. Failed logins in last 24h
    const failedLogins = await LoginHistory.count({
      where: {
        status: 'failed',
        created_at: { [Op.gte]: new Date(Date.now() - 24 * 60 * 60 * 1000) },
      },
      include: [{ model: User, as: 'user', where: { company_id: companyId }, attributes: [] }],
    });
    if (failedLogins > 10) {
      const weight = Math.min(30, failedLogins * 2);
      factors.push({ factor: 'failed_logins', weight, detail: `${failedLogins} failed logins in 24h` });
      totalScore += weight;
    }

    // 2. Inactive users ratio
    const totalUsers = await User.count({ where: { company_id: companyId } });
    const inactiveUsers = await User.count({ where: { company_id: companyId, is_active: false } });
    if (totalUsers > 0) {
      const ratio = inactiveUsers / totalUsers;
      if (ratio > 0.5) {
        const weight = Math.round(ratio * 20);
        factors.push({ factor: 'high_inactive_ratio', weight, detail: `${inactiveUsers}/${totalUsers} users inactive` });
        totalScore += weight;
      }
    }

    // 3. Locked accounts
    const lockedUsers = await User.count({
      where: { company_id: companyId, locked_until: { [Op.gt]: new Date() } },
    });
    if (lockedUsers > 0) {
      const weight = lockedUsers * 10;
      factors.push({ factor: 'locked_accounts', weight: Math.min(weight, 25), detail: `${lockedUsers} accounts currently locked` });
      totalScore += Math.min(weight, 25);
    }

    // 4. No admin activity in 30 days
    const company = await Company.findByPk(companyId);
    if (company?.last_activity_at) {
      const daysSince = (Date.now() - new Date(company.last_activity_at).getTime()) / (1000 * 60 * 60 * 24);
      if (daysSince > 30) {
        factors.push({ factor: 'admin_inactivity', weight: 15, detail: `No admin activity for ${Math.round(daysSince)} days` });
        totalScore += 15;
      }
    }

    // 5. Mass data operations (many audit logs for exports/deletes in 24h)
    const massOpsWhere = {
      action: { [Op.iLike]: '%export%' },
      created_at: { [Op.gte]: new Date(Date.now() - 24 * 60 * 60 * 1000) },
    };
    if (companyId) massOpsWhere.company_id = companyId;
    const massOps = await AuditLog.count({ where: massOpsWhere });
    if (massOps > 20) {
      factors.push({ factor: 'mass_data_operations', weight: 20, detail: `${massOps} export operations in 24h` });
      totalScore += 20;
    }

    // Clamp to 100
    totalScore = Math.min(totalScore, 100);
    const level = totalScore >= 75 ? 'critical' : totalScore >= 50 ? 'high' : totalScore >= 25 ? 'medium' : 'low';

    // Upsert risk score
    const [riskScore] = await RiskScore.findOrCreate({
      where: { entity_type: 'company', entity_id: companyId },
      defaults: { score: totalScore, level, factors, last_calculated_at: new Date(), company_id: companyId },
    });

    if (riskScore) {
      await riskScore.update({ score: totalScore, level, factors, last_calculated_at: new Date(), company_id: companyId });
    }

    return { score: totalScore, level, factors };
  }

  /**
   * Calculate risk for a user
   */
  async calculateUserRisk(userId) {
    const factors = [];
    let totalScore = 0;

    const user = await User.findByPk(userId);
    if (!user) throw new Error('User not found');

    // 1. Failed login attempts
    if (user.failed_login_attempts > 3) {
      const weight = Math.min(30, user.failed_login_attempts * 5);
      factors.push({ factor: 'failed_logins', weight, detail: `${user.failed_login_attempts} failed attempts` });
      totalScore += weight;
    }

    // 2. Account locked
    if (user.locked_until && new Date(user.locked_until) > new Date()) {
      factors.push({ factor: 'account_locked', weight: 20, detail: 'Account is currently locked' });
      totalScore += 20;
    }

    // 3. Multiple active sessions
    const activeSessions = await Session.count({
      where: { user_id: userId, is_active: true, expires_at: { [Op.gt]: new Date() } },
    });
    if (activeSessions > 5) {
      factors.push({ factor: 'many_sessions', weight: 15, detail: `${activeSessions} active sessions` });
      totalScore += 15;
    }

    // 4. No 2FA enabled (for admins)
    if (['main_admin', 'admin'].includes(user.role) && !user.two_factor_enabled) {
      factors.push({ factor: 'no_2fa', weight: 10, detail: 'Admin without 2FA' });
      totalScore += 10;
    }

    // 5. No recent login (inactive >60 days)
    if (user.last_login) {
      const daysSince = (Date.now() - new Date(user.last_login).getTime()) / (1000 * 60 * 60 * 24);
      if (daysSince > 60) {
        factors.push({ factor: 'long_inactive', weight: 15, detail: `No login for ${Math.round(daysSince)} days` });
        totalScore += 15;
      }
    }

    totalScore = Math.min(totalScore, 100);
    const level = totalScore >= 75 ? 'critical' : totalScore >= 50 ? 'high' : totalScore >= 25 ? 'medium' : 'low';

    const [riskScore] = await RiskScore.findOrCreate({
      where: { entity_type: 'user', entity_id: userId },
      defaults: { score: totalScore, level, factors, last_calculated_at: new Date(), company_id: user.company_id },
    });

    if (riskScore) {
      await riskScore.update({ score: totalScore, level, factors, last_calculated_at: new Date(), company_id: user.company_id });
    }

    return { score: totalScore, level, factors };
  }

  /**
   * Get all risk scores (for dashboard)
   */
  async getAllRiskScores({ entityType, minScore = 0, page = 1, limit = 20, companyId } = {}) {
    const where = {};
    if (entityType) where.entity_type = entityType;
    if (minScore > 0) where.score = { [Op.gte]: minScore };
    if (companyId) where.company_id = companyId;

    const { count, rows } = await RiskScore.findAndCountAll({
      where,
      order: [['score', 'DESC']],
      limit,
      offset: (page - 1) * limit,
    });

    return { scores: rows, total: count, page, pages: Math.ceil(count / limit) };
  }

  /**
   * Recalculate all company risks
   */
  async recalculateAll() {
    const companies = await Company.findAll({ where: { is_active: true }, attributes: ['id'] });
    const results = [];
    for (const company of companies) {
      try {
        const risk = await this.calculateCompanyRisk(company.id);
        results.push({ company_id: company.id, ...risk });
      } catch (e) {
        results.push({ company_id: company.id, error: e.message });
      }
    }
    return results;
  }

  /**
   * Get risk alerts (high/critical scores)
   */
  async getAlerts(companyId) {
    const where = { level: { [Op.in]: ['high', 'critical'] } };
    if (companyId) where.company_id = companyId;
    const alerts = await RiskScore.findAll({
      where,
      order: [['score', 'DESC']],
      limit: 50,
    });
    return alerts;
  }
}

module.exports = new RiskService();
